src/parser.c

```diff
@@ -0,0 +1,122 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <assert.h>
+#include "parser.h"
+
+#define MAXCHAR 4098
+#define INITLINE 100
+#define READ_BUFFERSIZE 256
+#define LSH_TOK_BUFFERSIZE 64
+#define LSH_TOK_DELIM " \t\r\n\a"
+
+/* Parsing Functions for Parsing Input Files */
+
+/* returns with pointer to array of strings, each of which represent a line in a given file */
+/* current limit to each string is 1000 characters */
+/* return NULL if file could not be opened */
+char** parse_file(char* filename)
+{
+	FILE *f = fopen(filename, "r");
+
+	if (f == NULL) {
+		return NULL;
+	}
+
+    // initialization
+	int n = 0;
+	int i = INITLINE;
+	char str[MAXCHAR];
+	char** lines;
+	lines = calloc(i, sizeof(char*));
+	if (lines == NULL) {
+		fprintf(stderr, "lineparse_file: calloc failed\n");
+		exit(0);
+	}
+
+    // read each line into a string. This size is currently limited at 1025 lines (1024, including terminating char).
+    // resize if necessary while reading file
+	while(fgets(str, MAXCHAR, f) != 0) {
+    if (n >= i) {
+      i += i;
+      lines = realloc(lines, i * sizeof (char*));
+      if (lines == NULL) {
+        fprintf(stderr, "lineparse_file: realloc failed\n");
+        exit(0);
+      }
+    }
+
+		lines[n] = strdup(str);
+		n++;
+	}
+  lines[n] = NULL;
+
+	fclose(f);
+	return lines;
+}
+
+/* Parsing functions for parsing command line inputs */
+/* read a command line and return a string */
+char* parse_read_line()
+{
+	char* input = calloc(READ_BUFFERSIZE, sizeof(char*));
```

Comment:
This calloc is wrong, you're allocating too much memory. From `man calloc`:

> The  calloc()  function allocates memory for an array of nmemb elements of size bytes each and returns a pointer to the allocated memory.

You want `sizeof(char)`.

Reply:
calloc for this has been fixed.